### PR TITLE
[Thread Dump Analyzer Tool] Restructure the backend, add structured logging, fail-fast uploads, per-file gzip, and a security-hardening pass

### PR DESCRIPTION
**Pull request by msakiththiyan**
## Purpose

Following the previous PR (which moved TDAT into `apps/`), this PR restructures the backend from a flat package into a layered `cmd/api` plus `internal/{config,job,logger,transport/http}` layout, adds structured logging, makes uploads fail fast on bad files, and gzips uploads per file so large multi-file dumps clear gateway request-size caps. It then layers on a security-hardening pass (AI-prompt data scrubbing, per-user result isolation, an AI opt-out, browser security headers, request-wide memory caps, and self-contained SCA tooling) and replaces the transient upload Snackbar with a notification system that surfaces the backend's real error. No API contract change.

## Goals

- Move the backend to an idiomatic layout (`cmd/` entrypoint, `internal/` packages by concern) and add tests at the package boundary.
- Give operators a readable console log plus an optional on-disk session log, controlled by `LOG_LEVEL` and `LOG_FILE`, with the container defaulting to stdout/stderr.
- Reject a bad upload up front (0 parsed threads, or 0 parsed usage rows) and mark the job `failed` with the file name, instead of a misleading `completed`.
- Let a ~54 MB multi-file upload reach the backend through a gateway that caps request bodies, without per-platform config and without changing the parser.
- Parse real-world `top`/usage files (header rows, `LWP`/`SPID` synonyms, decimal `nid` from JDK 21+) and keep the AI call from emitting a misleading parse error on truncation.
- Bound a single request's memory: a per-request decompressed-bytes total and a per-field file-count cap, on top of the existing per-file inflate ceiling.
- Keep customer data out of the AI prompt (scrub secrets/PII), let an operator turn the AI call off entirely, reject tokens minted for another Asgardeo app, and bind each job's result to its creator.
- Ship browser security headers and a self-contained SCA script so the app gates its own dependency risk without depending on a repo-root `.github/`.
- Surface the backend's real failure reason to the user in a centered modal, instead of a 4-second Snackbar that hid the cause.

## Approach

A series of self-contained commits, grouped by theme.

**1. Backend package restructure**

- Relocates the entrypoint to `cmd/api/main.go` (`go build ./cmd/api`).
- Splits the flat package into `internal/`: `config/config.go` (`Config` + `LoadConfig`), `job/service.go` (`Job`, `JobStore`, `JobLimiter`, the `runAnalysis` pipeline, `AggregatedAnalysisResponse`), and `transport/http/` (`handler.go` for `NewRouter` and the analyze/job-status handlers, `middleware.go` for Bearer-JWT auth and the per-IP `IPLimiter`).
- Adds `internal/job/service_test.go` and `internal/transport/http/{handler,middleware}_test.go`; removes the old root `limiters_test.go` (coverage moved into the middleware test).
- No request/response contract change.

**2. Structured logging**

- New `internal/logger/logger.go`: a `tint` console handler (colorized only on a TTY) plus an optional plain-text session file, combined through a small `fanoutHandler`.
- `LOG_LEVEL` and `LOG_FILE` are the only knobs. Empty `LOG_FILE` writes a timestamped `logs/tdat-session-<ts>.log`; `off`/`none`/`false`/`-` disables it; a path pins it. The Docker image sets `LOG_FILE=off`.

**3. Fail-fast upload validation**

- `runAnalysis` now returns `(*AggregatedAnalysisResponse, error)`. A non-dump file (0 threads) or malformed usage file (0 rows) yields a `*validationError` naming the file; the pipeline aborts after phase 1, so no rules, health score, or AI call runs on a bad batch.
- `api/analyze.ts` distinguishes a network/gateway cut-off from an HTTP error and surfaces an actionable message. `READ_TIMEOUT`/`WRITE_TIMEOUT` defaults raised to 10m for large uploads.

**4. Per-file gzip upload**

- The frontend gzips each file with `CompressionStream` and appends it as a normal multipart part, preserving order; browsers without `CompressionStream` fall back to raw upload.
- The backend sniffs the gzip signature (`0x1f 0x8b`) in `readMultipartFiles` and inflates each part through a size-capped reader, so the parser stays encoding-agnostic and raw uploads still work.
- The frontend compresses one file at a time (not all blobs at once) so peak browser memory stays bounded on large batches.
- `MAX_UPLOAD_BYTES` bounds the compressed wire body; a new `MAX_DECOMPRESSED_BYTES` (default 2x, 200MB) caps each inflated part against decompression bombs, and `MAX_TOTAL_DECOMPRESSED_BYTES` (default 2x that, 400MB) caps the running total across all parts (dumps and usages combined) so many small parts can't sum into an OOM.

**5. Usage parsing and AI robustness**

- `internal/parser/thread_parser.go` is header-aware: maps columns by name (`TID`/`LWP`/`SPID`, `%CPU`/`PCPU`/`CPU`, optional `PID`/`TIME`) and falls back to fixed positions when headerless. Native IDs and usage TIDs parse as hex or decimal, so JDK 21+ decimal `nid` correlates.
- `internal/ai/insights.go` raises `MaxTokens` to 4096 and detects a `max_tokens` stop reason to fail cleanly instead of a misleading parse error; a bounded preview of the raw payload stays in the server log (see the security pass below).

**6. Security hardening**

- **AI-prompt data scrubbing**: new `internal/ai/scrub.go` redacts secrets and PII (auth headers, JWTs, `key=value` secrets, emails, UUIDs, IPv4) from thread names, issues, and stack frames before they enter the Anthropic prompt. Always on and prompt-only, so the result returned to the caller stays unredacted; `quoteAll` still handles prompt injection while `scrub` handles confidentiality. The `key=value` matcher redacts quoted values (single- or double-quoted) as a whole, so a secret containing spaces is not left partly exposed.
- **AI opt-out**: `AI_INSIGHTS_ENABLED` (default true) gates the outbound call; when false the pipeline keeps every thread in-process and returns a fixed "insights disabled by policy" payload instead of sending data to Anthropic.
- **Per-user result isolation**: `RequireAuth` now carries the validated `sub` in the request context; `JobStore.Create` records it as `OwnerSub`, and `jobStatusHandler` returns `404` (not `403`, so existence stays unconfirmable) when a different subject reads a job. `Config.Validate` fails boot when `AUTH_ENABLED=true` but `JWT_AUDIENCE` is unset, so a token minted for another Asgardeo app can't be replayed. A validated token with an empty `sub` is rejected (401): the empty owner is the auth-off sentinel, so an unattributable token must not be bindable to a job.
- **Audit logging**: job lifecycle events (created, completed, invalid upload, timed out, panicked, cross-owner denial) log `job_id` and the JWT `sub` via `slog`; nothing is persisted, so these structured lines are the audit trail.
- **Browser security headers**: a shared `frontend/security-headers.conf` (CSP, `X-Content-Type-Options`, `X-Frame-Options`, `Referrer-Policy`, `Permissions-Policy`, HSTS) is included at the nginx server level and re-included in the `/config.js` block, since nginx replaces rather than merges inherited `add_header`. `docker-entrypoint.sh` now escapes the runtime config values (backslashes, single quotes, and CR/LF) before writing `config.js`, so a hostile or malformed env value can't break out of the JS string literal.
- **Upload fan-out guard**: a `MAX_FILES_PER_REQUEST` per-field count cap (default 200) rejects oversized batches with 400 before any work, bounding the per-dump goroutine/parse fan-out; the handler also `defer`s `r.MultipartForm.RemoveAll()` so any temp files `ParseMultipartForm` spilled are always dropped.
- **SCA tooling**: a self-contained `scripts/sca.sh` plus `Makefile` targets (`make sca`/`sca-backend`/`sca-frontend`/`sca-trivy`) run govulncheck, `pnpm audit`, and Trivy. Only `pnpm audit` high+ fails the run; govulncheck and Trivy are report-only, and a scanner that can't run (missing tool, failed install, or a non-report exit code) is surfaced as an incomplete signal rather than a clean pass. No repo-root `.github/` is touched.

**7. Frontend notification system**

- New `context/NotificationContext.tsx` provides a `useNotify(message, severity?, options?)` hook backed by a centered MUI `Dialog`: alerts queue and show one at a time, backdrop clicks are ignored so a message can't be dismissed before it is read, and severity drives the icon, accent colour, and default title.
- New `utils/jobError.ts#describeJobError` maps a failed job's backend error (the same text the server logs) into a user-facing message: a generic `internal error (ref=...)` keeps only the reference id, while validation/timeout/parse reasons (joined by `; `) are shown one per line.
- `pages/upload/index.tsx` drops the transient 4-second `Snackbar` and the inline `Alert`; the upload, duplicate-skip, no-threads, and job-failed paths all route through `notify`, so the user now sees the backend's real reason instead of a generic "analysis failed".

**8. Toolchain and hygiene**

- Bumps Go to 1.25.0 in `go.mod`/`go.sum` and the backend Dockerfile base image.
- Fixes naming inconsistencies across the READMEs and Dockerfile, and a trailing-newline fix in the tool `.gitignore`.

## User stories

- **As a Customer Success engineer**, I can upload large multi-file bundles without running into request size limits because files are compressed in the browser and expanded on the server.
- **As a Customer Success engineer**, I get immediate, actionable errors that identify the failing file and can enable session logging with `LOG_FILE` for easier troubleshooting.
- **As a contributor**, I can navigate and extend the backend more easily thanks to a standard Go project structure and package-level tests.
- **As a security reviewer**, customer secrets and PII are scrubbed before any thread text leaves for the AI API, the whole AI call can be switched off, each job's result is readable only by the subject that created it, and `make sca` scans dependencies, so I can sign off without an external CI dependency.
- **As a user whose upload fails**, I now see the backend's actual reason (a named bad file, a timeout, or a reference id) in a centered alert I have to dismiss, instead of a 4-second toast that disappears before I read it.

## Release note

> Backend reshaped into `cmd/api` plus `internal/{config,job,logger,transport/http}` with package-level tests; no API contract change. Adds structured logging with an optional session-file mirror (`LOG_LEVEL`/`LOG_FILE`), fail-fast upload validation that names the bad file, and per-file gzip on upload (browser compresses, backend inflates under `MAX_DECOMPRESSED_BYTES`) so large multi-file uploads clear gateway request-size caps. Hardens the app: secrets/PII are scrubbed from the AI prompt, the AI call can be disabled (`AI_INSIGHTS_ENABLED`), results are isolated per JWT subject with a required `JWT_AUDIENCE`, requests are bounded by per-request decompressed-byte and file-count caps, nginx serves CSP/security headers, and `make sca` runs govulncheck/`pnpm audit`/Trivy. The upload page now surfaces the backend's real error in a modal. Also improves usage-file parsing and the AI insights call, and bumps the toolchain to Go 1.25.

## Documentation

- `backend/README.md`, root `README.md`, `frontend/README.md` - logging knobs (`LOG_LEVEL`/`LOG_FILE`), gzip-on-upload, the upload caps (`MAX_UPLOAD_BYTES`, `MAX_DECOMPRESSED_BYTES`, `MAX_TOTAL_DECOMPRESSED_BYTES`, `MAX_FILES_PER_REQUEST`), the `AI_INSIGHTS_ENABLED` opt-out, `JWT_AUDIENCE`, the browser security headers, and the SCA targets.
- `backend/openapi.yaml` - per-part gzip detection, the size and file-count caps, the header-aware `thread_usages` format, the cross-owner `404`, and refined 400/429 error descriptions.
- `backend/.env.example`, root `.env.example` - the upload caps, logging knobs, `AI_INSIGHTS_ENABLED`, and the now-required `JWT_AUDIENCE`.
- `CONTRIBUTING.md` - the `make sca` workflow for local dependency scanning.

## Training

N/A - internal tool. No external training material updates required.

## Certification

N/A - TDAT is a developer/support tool and is not part of any product certification track.

## Marketing

N/A - internal tool. No marketing collateral planned.

## Automation tests

- New package-level tests cover the restructured layout: `internal/job/service_test.go` and `internal/transport/http/{handler,middleware}_test.go` (the latter absorbs the removed root `limiters_test.go`). `internal/parser/thread_parser_test.go` gains header-aware and decimal-NID cases.
- The security pass adds `internal/ai/scrub_test.go` (redaction cases, incl. quoted-value secrets) and `internal/config/config_test.go` (the `Validate` audience invariant); `handler_test.go` and `middleware_test.go` gain cross-owner-read (404), empty-subject rejection, per-request total-cap, and file-count-cap cases.
- Run the full suite with `go test ./...` from `backend/`.
- `make sca` runs the dependency scans locally (govulncheck, `pnpm audit`, Trivy); only `pnpm audit` high+ gates the run.
- Pre-merge note: the gzip fix assumes the gateway measures the compressed wire bytes, not the inflated content; worth a quick confirmation before leaning on it in production.

## Security checks

- No secrets committed; `.env`/`.env.*` remain gitignored with only `.env.example` allowlisted.
- **Outbound data minimization**: `internal/ai/scrub.go` redacts secrets and PII (auth headers, JWTs, `key=value` secrets, emails, UUIDs, IPv4) from thread text before it enters the AI prompt; `AI_INSIGHTS_ENABLED=false` keeps all thread data in-process. The AI JSON-parse failure now logs a bounded 256-char preview, not the full payload.
- **Per-user isolation**: jobs are bound to the creator's JWT `sub`, so cross-owner reads get `404`. `Config.Validate` refuses to boot with `AUTH_ENABLED=true` and no `JWT_AUDIENCE`, so a token minted for another Asgardeo app is rejected, as is a validated token carrying no `sub` (it can't be attributed to an owner). Job lifecycle events are audit-logged with the `sub`.
- **Upload memory safety**: the per-file `MAX_DECOMPRESSED_BYTES` ceiling (default 200MB) now sits under a per-request `MAX_TOTAL_DECOMPRESSED_BYTES` total (default 400MB) and a `MAX_FILES_PER_REQUEST` count cap (default 200), with `MAX_UPLOAD_BYTES` still capping the compressed body and `r.MultipartForm.RemoveAll()` dropping spilled temp files.
- **Browser hardening**: nginx serves a CSP plus `X-Content-Type-Options`, `X-Frame-Options`, `Referrer-Policy`, `Permissions-Policy`, and HSTS from `security-headers.conf`; `docker-entrypoint.sh` escapes runtime config values (quotes and CR/LF) before writing `config.js`.
- **Dependency scanning**: `scripts/sca.sh` (`make sca`) runs govulncheck, `pnpm audit` (gates on high+), and Trivy, kept app-local with no repo-root `.github/`; a scanner that fails to run is reported as an incomplete signal, never a silent pass.
- JWT auth and per-IP/per-instance rate limiting still live in `internal/transport/http/middleware.go`; auth now additionally enforces the audience claim and carries the subject downstream.
- Both images still run rootless; the backend sets `LOG_FILE=off` so it never writes a session log to a read-only filesystem.

## Samples

- `docker compose up --build` brings up the full stack (frontend `:8081`, backend `:8080`).
- The backend still serves the built-in HTML upload form at `GET /`; raw uploads work there via the same gzip sniff.
- `LOG_FILE=logs/run.log go run ./cmd/api` mirrors the console log to a file for a local session.
- `make sca` runs the dependency scans (govulncheck, `pnpm audit`, Trivy) from the app root.

## Related PRs

- https://github.com/wso2-open-operations/common-tools/pull/24

## Migrations (if applicable)

- Build path changed: the entrypoint moved to `cmd/api`, so build with `go build -o bin/tdat-backend ./cmd/api` and run with `go run ./cmd/api`. The in-repo Dockerfile is already updated. No data or schema migration; the HTTP contract is unchanged.
- New required env: `JWT_AUDIENCE` must be set wherever `AUTH_ENABLED=true` (the default) or the backend now refuses to start; set it to the app's Asgardeo client ID. Local dev can instead set `AUTH_ENABLED=false`.

## Test environment

- **Backend runtime** - Go 1.25+ toolchain on Linux; backend image base bumped to `golang:1.25.0-alpine`.
- **Containers** - Docker with Docker Compose; images verified rootless with a runtime-injected port and `LOG_FILE=off`.
- **Frontend** - Node 20 LTS, pnpm 9.x; `CompressionStream` required for gzip upload, with raw fallback otherwise.
- **Dependency scanning** - `scripts/sca.sh` shells out to govulncheck (auto-installed via `go install` if absent), `pnpm audit`, and Trivy (PATH binary, else the `aquasec/trivy` image, else skipped).

## Learning

- **A flat package hides seams; `internal/` by concern exposes them**: separating config, job orchestration, logging, and HTTP made the handler/middleware boundary testable in isolation.
- **Compress on the client, decompress at the boundary**: gzipping each file and inflating it in `readMultipartFiles` clears gateway caps without a per-platform size threshold and keeps the parser encoding-agnostic. Whole-body `Content-Encoding: gzip` was rejected because intermediaries handle it inconsistently.
- **Fail fast beats a misleading success**: returning `(*resp, error)` and aborting after phase 1 turns a bad upload into a `failed` job that names the file.
- **Truncation is not a parse error**: checking the `max_tokens` stop reason before unmarshalling lets the AI call report a budget overflow instead of a confusing JSON parse failure.
- **Scrub for confidentiality, quote for injection**: keeping the two passes separate lets `scrub` redact secrets/PII from the prompt while leaving the caller's result unredacted, and `quoteAll` independently neutralizes prompt injection.
- **nginx replaces inherited `add_header`, never merges**: any block that sets its own header drops the server-level ones, so `security-headers.conf` is re-included in the `/config.js` location rather than relying on inheritance.
- **Validate config before boot, not per request**: a cross-field `Config.Validate` (auth requires an audience) fails the process at startup instead of letting a misconfigured deploy accept replayable tokens.
- **An empty subject is not a valid owner**: auth-off mode stores an empty `OwnerSub`, so a validated token carrying no `sub` must be rejected rather than bound to a job, or it would silently inherit the auth-off (public) access path.
- **Exit code 3 is a result; any other non-zero is a failure**: `govulncheck` exits 3 when it finds vulnerabilities, so treating every non-zero as "findings reported" would hide a scanner that never actually ran. The script distinguishes the two and flags an incomplete scan.

## Release Notes

* **New Features**
  * Frontend now gzip-compresses uploads when supported; backend inflates gzip inputs with configurable per-file, per-request, and file-count limits.
  * More resilient `thread_usages` parsing (header-aware mapping, decimal/hex TID, better correlation) and enhanced job reporting (Native ID, clearer async failures).
  * Optional session file logging via `LOG_FILE`, plus improved dependency/security scanning guidance.

* **Bug Fixes**
  * Fail-fast handling for malformed inputs prevents partial results; job ownership enforcement returns 404; auth-enabled startup now requires `JWT_AUDIENCE`.

* **Documentation**
  * Updated quickstart, Docker/env instructions, API/OpenAPI details, and security header behavior.

### CODE REVIEW

File: apps/thread-dump-analyzer-tool/backend/internal/ai/insights.go [hunk 1]
```diff
@@ -95,7 +104,9 @@ func GetInsights(parentCtx context.Context, threads []analyzer.AnalyzedThread, u
 
 	var insights AIInsights
 	if err := json.Unmarshal([]byte(raw), &insights); err != nil {
-		return nil, fmt.Errorf("failed to parse JSON insights: %w (raw=%s)", err, raw)
+		// Keep the raw payload in the server log, out of the user-facing error.
+		slog.Warn("AI insights JSON parse failed", "error", err, "raw", raw)
+		return nil, fmt.Errorf("AI summary was not valid JSON")
```
[the remark this answers is not in the document]
Comment (msakiththiyan) [reply apps/thread-dump-analyzer-tool/backend/internal/ai/insights.go]: Done. `insights.go` now logs a 256-char preview plus `raw_len` instead of the full payload.

File: apps/thread-dump-analyzer-tool/backend/internal/transport/http/handler.go [hunk 2]
```diff
@@ -132,6 +69,118 @@ func NewRouter(cfg *Config, jobStore *JobStore, engine *analyzer.RuleEngine, enr
...
+		if err != nil {
+			return nil, fmt.Errorf("read %s: %w", h.Filename, err)
+		}
+		data, err = maybeGunzip(h.Filename, data, maxDecompressedBytes)
+		if err != nil {
+			return nil, err
+		}
+		out = append(out, job.FilePayload{FileName: strings.TrimSuffix(h.Filename, ".gz"), Data: data})
+	}
+	return out, nil
```
[the remark this answers is not in the document]
Comment (msakiththiyan) [reply apps/thread-dump-analyzer-tool/backend/internal/transport/http/handler.go]: Resolved. `readMultipartFiles` enforces a cumulative `maxTotalBytes` cap via a shared running total across both dumps and usages (`MAX_TOTAL_DECOMPRESSED_BYTES`, default 400MB), covered by `TestReadMultipartFiles_RejectsCumulativeOverflow.`

File: apps/thread-dump-analyzer-tool/backend/internal/transport/http/handler.go [hunk 3]
```diff
@@ -132,6 +69,118 @@ func NewRouter(cfg *Config, jobStore *JobStore, engine *analyzer.RuleEngine, enr
...
+	// Cap body size so ParseMultipartForm keeps file parts in RAM, not $TMPDIR.
+	r.Body = http.MaxBytesReader(w, r.Body, maxUploadBytes)
+	if err := r.ParseMultipartForm(maxUploadBytes); err != nil {
+		slog.Warn("multipart parse failed", "request_id", reqID, "error", err)
+		http.Error(w, fmt.Sprintf("Could not process upload. Limit is %d MiB (ref=%s)", maxUploadBytes>>20, reqID), http.StatusBadRequest)
+		return
+	}
+
+	dumpHeaders := r.MultipartForm.File["thread_dumps"]
+	usageHeaders := r.MultipartForm.File["thread_usages"]
```
[the remark this answers is not in the document]
Comment (msakiththiyan) [reply apps/thread-dump-analyzer-tool/backend/internal/transport/http/handler.go:140]: Done. Added `defer r.MultipartForm.RemoveAll()` immediately after the `ParseMultipartForm` block.

File: apps/thread-dump-analyzer-tool/backend/internal/transport/http/middleware_test.go [hunk 4]
```diff
@@ -245,3 +248,92 @@ func TestNewAuthenticator_FailsFastWhenUnconfigured(t *testing.T) {
...
+		var captured string
+		h := l.limitByIP(func(w http.ResponseWriter, req *http.Request) { captured = req.RemoteAddr })
+		w := httptest.NewRecorder()
+		h(w, r)
+		return captured
+	}(r)
+	if !strings.HasPrefix(host, "10.0.0.5:") {
+		t.Errorf("RemoteAddr modified unexpectedly: %q", host)
+	}
+}
```
[the remark this answers is not in the document]
Comment (msakiththiyan) [reply apps/thread-dump-analyzer-tool/backend/internal/transport/http/middleware_test.go]: Resolved. Asserted IPLimiter bucket keying directly (port-stripped host, ignores X-Forwarded-For) and explicit multipart temp-file cleanup in the upload handler.

File: apps/thread-dump-analyzer-tool/frontend/src/api/analyze.ts [hunk 5]
```diff
@@ -26,20 +26,34 @@ const authHeader = async (getAccessToken: TokenGetter): Promise<HeadersInit> =>
...
+};
+
+// appendFiles gzips each file and appends it under field, preserving order so dump/usage indexes stay aligned.
+const appendFiles = async (formData: FormData, field: string, files: File[]): Promise<void> => {
+  if (!gzipSupported) {
+    files.forEach((file) => formData.append(field, file));
+    return;
+  }
+  const blobs = await Promise.all(files.map(gzipFile));
+  blobs.forEach((blob, i) => formData.append(field, blob, `${files[i].name}.gz`));
```
[the remark this answers is not in the document]
Comment (msakiththiyan) [reply apps/thread-dump-analyzer-tool/frontend/src/api/analyze.ts]: Done. `analyze.ts` now compresses and appends sequentially, so peak memory holds one gzip blob at a time instead of every compressed file at once.

File: apps/thread-dump-analyzer-tool/backend/internal/ai/scrub.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+// under the License.
+
+package ai
+
+import "regexp"
+
+// Secrets and PII scrubbed from thread text before it leaves for the AI API; specific patterns run before generic ones.
+var (
+	reAuth     = regexp.MustCompile(`(?i)\b(?:bearer|basic)\s+[A-Za-z0-9._~+/=-]+`)
+	reSecretKV = regexp.MustCompile(`(?i)\b([\w-]*(?:password|passwd|pwd|secret|token|api[_-]?key|access[_-]?key|private[_-]?key|credential|authorization)[\w-]*)(\s*[:=]\s*)[^\s,;"'\[]+`)
```
[the remark this answers is not in the document]
Comment (msakiththiyan) [reply apps/thread-dump-analyzer-tool/backend/internal/ai/scrub.go]: Fixed `reSecretKV` to redact quoted secret values (`password="hunter2"`, `access_token='abc'`) which previously leaked through unredacted, and added regression tests for both quoted variants.

File: apps/thread-dump-analyzer-tool/backend/internal/transport/http/middleware.go [hunk 7]
```diff
@@ -82,16 +157,30 @@ func (a *Authenticator) RequireAuth(next http.HandlerFunc) http.HandlerFunc {
...
+		if err != nil {
 			slog.Warn("rejected request: token validation failed", "error", err, "remote_addr", r.RemoteAddr, "path", r.URL.Path)
 			unauthorized(w, "invalid or expired token")
 			return
 		}
 
-		next.ServeHTTP(w, r)
+		// Carry the subject so handlers can bind a job to its creator and reject cross-user reads.
+		ctx := context.WithValue(r.Context(), subjectContextKey, tok.Subject())
+		next.ServeHTTP(w, r.WithContext(ctx))
```
[the remark this answers is not in the document]
Comment (msakiththiyan) [reply apps/thread-dump-analyzer-tool/backend/internal/transport/http/middleware.go:176]: Done. Added a `missing subject `test case that strips the `sub` claim from an otherwise-valid token and asserts `RequireAuth` rejects it with a 401 without running the protected handler.

File: apps/thread-dump-analyzer-tool/frontend/docker-entrypoint.sh [hunk 8]
```diff
@@ -21,11 +21,19 @@ set -e
...
+ASGARDEO_CLIENT_ID_ESC=$(escape_js "$ASGARDEO_CLIENT_ID")
+ASGARDEO_BASE_URL_ESC=$(escape_js "$ASGARDEO_BASE_URL")
 cat > /usr/share/nginx/html/config.js <<EOF
 window.configs = {
-    apiUrl: '${API_URL}',
-    ASGARDEO_CLIENT_ID: '${ASGARDEO_CLIENT_ID}',
-    ASGARDEO_BASE_URL: '${ASGARDEO_BASE_URL}'
+    apiUrl: '${API_URL_ESC}',
+    ASGARDEO_CLIENT_ID: '${ASGARDEO_CLIENT_ID_ESC}',
+    ASGARDEO_BASE_URL: '${ASGARDEO_BASE_URL_ESC}'
```
[the remark this answers is not in the document]
Comment (msakiththiyan) [reply apps/thread-dump-analyzer-tool/frontend/docker-entrypoint.sh:41]: Done. Hardened `escape_js` to also escape CR/LF, using a guarded `$!{N;ba}` slurp that, unlike the naive `:a;N;$!ba`, still escapes single-line values correctly

File: apps/thread-dump-analyzer-tool/scripts/sca.sh [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+  elif [ -x "$(go env GOPATH)/bin/govulncheck" ]; then
+    bin="$(go env GOPATH)/bin/govulncheck"
+  else
+    echo "    installing govulncheck..."
+    go install golang.org/x/vuln/cmd/govulncheck@latest
+    bin="$(go env GOPATH)/bin/govulncheck"
+  fi
+  # Reachability-aware: only flags vulns in code paths actually called.
+  ( cd "$BACKEND_DIR" && "$bin" ./... ) || echo "    findings reported (non-blocking)"
+}
```
[the remark this answers is not in the document]
Comment (msakiththiyan) [reply apps/thread-dump-analyzer-tool/scripts/sca.sh:73]: Resolved. Reworked `run_govulncheck` to stop masking install/execution failures as findings: it now skips cleanly if install or the binary is missing, and branches on the exit code so only exit 3 reports "findings (non-blocking)" while any other non-zero prints a distinct execution-failure WARN.

**Review by v15a1 — APPROVED** [review]
